### ISSUE TITLE
Hints for Youtube Fling

### ISSUE THREAD

**Issue by liuche** [mozilla-mobile/firefox-tv#1286]
### Why/User Benefit/User Problem
After we add #1284 users will be able to cast youtube videos from the Youtube app (but not generic links), but if they didn't already know about this feature, how will they learn about it?

### What / Requirements
Provide a hint so users know about this capability.

@cartertrout do you know what kind of branding requirements we'd have, if this feature is only to highlight "Youtube casting"? Could you suggest on getting Youtube branding, casting icon, etc?

### Acceptance Criteria (how do I know when I’m done?)
* Hint added so users know about casting to Firefox with Youtube.
* Probe added to keep track of how many times the hint is displayed

**Comment by aminalhazwani** [comment]
⚠️ **IN PROGRESS** ⚠️ 

I was able to cast from the YouTube mobile and tablet app to the Fire TV via Silk. I wasn't able to cast from Laptop/Chrome to the Fire TV. For now we could only focus on casting from mobile/tablet YouTube app to Firefox.

If you tap the cast icon in the homescreen of the YouTube app YouTube tells you that there are no videos in the queue:

<img width="2022" alt="image" src="https://user-images.githubusercontent.com/1721875/47025481-de849b00-d163-11e8-9400-2ba18b9b7ed6.png">

then Silk opens and it opens youtube.com and no video is played.

If you do the same process with a Chromecast/Android TV this message is displayed on the TV:

![img_0649](https://user-images.githubusercontent.com/1721875/47025850-8a2deb00-d164-11e8-9598-02945f811181.jpg)

First question **1) can we display a custom message if users simply tap the cast icon in the YouTube app homescreen?**

In order to notify users that this feature is available we can display a small banner/message. The banner will tell users that they can cast videos from their mobile or tablet YouTube app. They would simply have to tap the cast icon on their device and select their Fire TV from the list.

### When can we show this banner

- on launch after update/install, similarly as we do for the Turbo Mode onboarding
  - this is unfortunately not very contextual, user action-related, we might want to avoid this for such specific feature
- people launching youtube.com/tv from the homescreen
  - **question 2) does it go against our policy?**
- people typing youtube.com from the homescreen
  - **question 2) does it go against our policy?**
- people typing on youtube.com/tv (youtube uses a custom keyboard)
  - **question 2) does it go against our policy?**
- people coming from YouTube.com app
  - **question 3) do we know when external apps open Firefox?**

### How can this banner look

<img width="1192" alt="image" src="https://user-images.githubusercontent.com/1721875/47031791-f878aa80-d170-11e8-9b2f-c417c0cc9a9e.png">

A banner that shows somewhere (see questions above), fixed to the bottom of the TV, with the button focused (like the pocket onboarding) ready to be dismissed.

Worth considering: if we display the message not on website we don't need to brand the message, we simply expose the feature. But if we display the message while users browse we might want to brand it since it might be not immediately clear that the message comes from Firefox and not from the website.

### What we need to tell

cc @BrianNJones: People can access this feature via the YouTube app on their mobile or tablet device. They need to tap the cast icon and then select their Fire TV from the list of shown devices. Something like:

Text: “You can now cast YouTube videos to Firefox for Fire TV. Tap the cast icon from your YouTube mobile or tablet app.”

Button: “Ok, got it”

### What we would like to measure

- Do people cast after seeing the banner/message? 
- Others TBD

**Comment by aminalhazwani** [comment]
Some early explorations...

<img width="1306" alt="image" src="https://user-images.githubusercontent.com/1721875/47093821-46051e00-d22a-11e8-8b22-209e9e42a21c.png">

We could slide the banner from the bottom of the screen so that motion can drive attention and enable users to understand where the current focus is.

**Comment by liuche** [comment]
@BrianNJones this will be in our next sprint so we'll also want strings for this - assigning you.

**Comment by BrianNJones** [comment]
@liuche @aminalhazwani , been trying to get this as short as possible, but there are quite a few "moving parts" in this little banner string:

Cast YouTube videos to Firefox for Fire TV from your phone or tablet. Open YouTube on your device and tap the <icon> to get started.

**Comment by BrianNJones** [comment]
@liuche @aminalhazwani, hi. trying to get this as short as possible, but there are several moving parts that need to get menitioned:

Cast YouTube videos to Firefox for Fire TV from your phone or tablet. To get started, tap [icon] in the YouTube app on your device.

**Comment by aminalhazwani** [comment]
Thanks @BrianNJones! Here how it looks like with the string

<img width="1941" alt="image" src="https://user-images.githubusercontent.com/1721875/47426102-916c7e80-d78c-11e8-8aa6-ec2e87a1200c.png">

@liuche as always I published the web inspector here https://github.com/mozilla-mobile/firefox-tv/wiki/Design#mocks-with-a-web-interface

For the time being the cast icon is available here https://github.com/FirefoxUX/photon-icons/pull/41

**Comment by BrianNJones** [comment]
@aminalhazwani @liuche , if we shortened the string to:
Cast YouTube videos from your phone or tabled. Open YouTube on your device and tap [icon] to get started.

...do you think it's obvious that we're talking about casting to Firefox for Fire TV?

**Comment by liuche** [comment]
@BrianNJones I think that seems reasonable - one thing that might not be represented here (that we need to test once we implement this) is that the user might need to select "Firefox or Silk" to actually complete the action. Should we say "Firefox"? or should that be obvious?

**Comment by BrianNJones** [comment]
@aminalhazwani @liuche , if we shortened the string to:
Cast YouTube videos from your phone or tablet. Open YouTube on your device and tap [icon] to get started.

...do you think it's obvious that we're talking about casting to Firefox for Fire TV?

**Comment by aminalhazwani** [comment]
@BrianNJones it depends on how the first interaction will look like -- Amazon Silk can cast too. Once we get this implemented we can see what the OS prompts to the users the first time they try to cast from their device. In the meantime this is how the new string would look like.

<img width="1174" alt="image" src="https://user-images.githubusercontent.com/1721875/47489180-4adb5a80-d846-11e8-8d83-02f0d9e27043.png">

**Comment by BrianNJones** [comment]
> 
> 
> @BrianNJones I think that seems reasonable - one thing that might not be represented here (that we need to test once we implement this) is that the user might need to select "Firefox or Silk" to actually complete the action. Should we say "Firefox"? or should that be obvious?

@liuche, Right! We'll need to reference Firefox directly, I think. I'm on board with what @aminalhazwani is thinking, though, to be sure: let's see how the OS prompt behaves and go from there.

**Comment by liuche** [comment]
After testing the new casting behavior, we've found that there actually...isn't a way to choose Firefox or Silk - the OS just decides somewhat arbitrarily.

Still waiting on @athomasmoz to see what we want to do about this Silk/Youtube casting issue. @athomasmoz if we don't figure out a way to cast directly to Firefox, will we move this out of this sprint?

For now, I'll just export these strings, because it doesn't look like there's any UI to select a choice.

**Comment by liuche** [comment]
We can use [ImageSpan](https://developer.android.com/reference/android/text/style/ImageSpan#ImageSpan(android.graphics.drawable.Drawable,%20int)) to embed the cast icon in the text.

**Comment by brampitoyo** [comment]
@athomasmoz @liuche Posting to clarify what Amin had designed a couple of months back.

Our desired behaviour for YouTube cast hint is as follows:

1. Many roads lead to YouTube:
    * Launch from FFTV homescreen tile
    * Launch from Fire TV home icon
    * Type youtube.com, youtube.com/tv, or any variant of that in the FFTV address bar
2. In all the above cases, when you first visit YouTube, show a splash screen/modal dialogue/toast
3. When you visit YouTube again, show a subtle hint in the hint bar

![youtube-casting-hint-behaviour](https://user-images.githubusercontent.com/51007/56180418-612bdb80-605d-11e9-9b69-6d622d6a93fc.png)

Amin had designed a toast to solve this problem. It was a good idea. However, since we’re only going to show this message once, we want to make sure that it’s noticed and read.

When a stronger visual is desired, we can go with either:

1. A modal dialogue, or
2. A splash screen

![youtube-casting-notification-styles@2x](https://user-images.githubusercontent.com/51007/56180499-afd97580-605d-11e9-8784-c675adeec93b.png)

Thoughts?

**Comment by liuche** [comment]
I like the modal dialog better, because it's more contextual (and we already have 2 other splash screens - Pocket and Turbo Mode).

Otherwise, looks good!

**Comment by brampitoyo** [comment]
@liuche Great. Let’s go with the modal dialogue, then.

#### Cast icon
https://cl.ly/712aac54d8ad/Cast.svg

#### Illustration
<img width="180" src="https://user-images.githubusercontent.com/51007/56339039-a8060680-6200-11e9-8b9a-fb7ced6166e2.png">

As far as design specs go, I’m happy using the standard modal dialogue that FFTV has (it seems that Silk Browser is able to invoke it).

**Comment by athomasmoz** [comment]
Closing as won't do after YouTube app announcement
